File: Parse/BptLoader.cpp
#include "Parse/BptLoader.hpp"

#include "Object/Impl/Shape/BezierPatch.hpp"
#include "Object/Impl/Shape/Group.hpp"
#include <fstream>

namespace Parse {
    std::unique_ptr<Object::Shape> BptLoader::load(const std::string &filename)
    {
        std::vector<std::unique_ptr<Object::Shape>> patches;

        std::ifstream file(filename.c_str());
        int numPatches;

        file >> numPatches;
        for (int i = 0; i < numPatches; i++) {
            int dimx, dimy;
            file >> dimx >> dimy;
            std::vector<Math::Point> controlPoints;
            for (int j = 0; j < 16; j++) {
                float x, y, z;
                file >> x >> y >> z;
                controlPoints.push_back(Math::Point(x, y, z));
            }
            patches.push_back(std::make_unique<Object::Impl::Shape::BezierPatch>(16, 16, std::move(controlPoints)));
        }

        if (patches.size() == 1) {
            return std::move(patches[0]);
        }
        else {
            return std::make_unique<Object::Impl::Shape::Group>(std::move(patches));
        }
    }
}

File: src/input/bpt_loader.rs
use crate::geo;
use geo::Point3;

use crate::object;
use object::Shape;

use object::shape::BezierPatch;
use object::shape::Group;

use std::fs::File;
use std::io::{BufRead, BufReader};
use std::str::FromStr;

pub struct BptLoader;

impl BptLoader {
    pub fn load(filename: &str) -> Option<Box<dyn Shape>> {
        if let Ok(file) = File::open(filename) {
            let mut reader = BufReader::new(file);

            let num_patches = Self::parse_nums::<i32, 1>(&mut reader)[0];

            let mut patches: Vec<Box<dyn Shape>> = Vec::new();
            for _ in 0..num_patches {
                let [_, _] = Self::parse_nums::<i32, 2>(&mut reader);
                let mut control_points = [Point3::ZERO; 16];
                for j in 0..16 {
                    let [x, y, z] = Self::parse_nums::<f32, 3>(&mut reader);
                    control_points[j] = Point3::new(x, y, z);
                }
                patches.push(Box::new(BezierPatch::new(
                    16,
                    16,
                    control_points,
                )));
            }

            return Some(Box::new(Group::new(patches)));
        } else {
            return None;
        }
    }

    fn parse_nums<T, const N: usize>(reader: &mut impl BufRead) -> [T; N]
    where
        T: FromStr + Copy + Default,
    {
        let mut line = String::new();
        let mut result = [T::default(); N];
        if reader.read_line(&mut line).is_ok() {
            for (i, part) in line.trim_end().split(' ').enumerate() {
                if i < N {
                    result[i] = part.parse().unwrap_or(T::default());
                }
            }
        }

        return result;
    }
}